src/vr/journal.zig

```diff
@@ -670,3 +1022,108 @@ pub const BitSet = struct {
         }
     }
 };
+
+/// Take a u6 to limit to 64 items max (2^6 = 64)
+pub fn IOPS(comptime T: type, comptime size: u6) type {
+    const Map = std.meta.Int(.unsigned, size);
+    const MapLog2 = std.math.Log2Int(Map);
+    return struct {
+        const Self = @This();
+
+        items: [size]T = undefined,
+        /// 1 bits are free items
+        free: Map = std.math.maxInt(Map),
+
+        pub fn acquire(self: *Self) ?*T {
+            const i = @ctz(Map, self.free);
+            assert(i <= @bitSizeOf(Map));
+            if (i == @bitSizeOf(Map)) return null;
+            self.free &= ~(@as(Map, 1) << @intCast(MapLog2, i));
+            return &self.items[i];
+        }
+
+        pub fn release(self: *Self, item: *T) void {
+            item.* = undefined;
+            const i = (@ptrToInt(item) - @ptrToInt(&self.items)) / @sizeOf(T);
+            assert(self.free & (@as(Map, 1) << @intCast(MapLog2, i)) == 0);
+            self.free |= (@as(Map, 1) << @intCast(MapLog2, i));
+        }
+
+        /// Returns true if there is at least one IOP available
+        pub fn available(self: *const Self) std.math.Log2IntCeil(Map) {
+            return @popCount(Map, self.free);
+        }
+
+        /// Returns true if there is at least one IOP in use
+        pub fn executing(self: *const Self) std.math.Log2IntCeil(Map) {
+            return @popCount(Map, std.math.maxInt(Map)) - @popCount(Map, self.free);
```

Comment:
Is this first `@popCount(Map, std.math.maxInt(Map))` exactly equal to our comptime `size`?

Reply:
The value is, but the type is not necessarily the same. `@popCount()` returns the smallest type that can hold all the possible values, which in this case is `std.math.Log2IntCeil(Map)`. Perhaps this would be clearer though:
```zig
 retrun math.maxInt(math.Log2IntCeil(Map)) - @popCount(Map, self.free);
 ```

Reply:
Ah yes, I was thinking only in terms of value and not type.

Rewriting the `maxInt()` to be over the function signature's result type is clearer.